radicle-cli/src/commands/inspect.rs

```diff
@@ -0,0 +1,222 @@
+                }
+                Long("payload") => {
+                    payload = true;
+                }
+                Long("history") => {
+                    history = true;
+                }
+                Long("id-only") => {
+                    id_only = true;
+                }
+                Value(val) if path.is_none() && id.is_none() => {
+                    let val = val.to_string_lossy();
+
+                    if let Ok(val) = Id::from_str(&val) {
+                        id = Some(val);
+                    } else if val.starts_with("rad:git:") {
+                        return Err(anyhow!("invalid ID '{}'", val));
+                    } else if let Ok(val) = PathBuf::from_str(&val) {
+                        path = Some(val);
+                    } else {
+                        return Err(anyhow!("invalid path or ID '{}'", val));
+                    }
+                }
+                _ => return Err(anyhow::anyhow!(arg.unexpected())),
+            }
+        }
+
+        Ok((
+            Options {
+                id,
+                path,
+                payload,
+                history,
+                refs,
+                id_only,
+            },
+            vec![],
+        ))
+    }
+}
+
+pub fn run(options: Options, ctx: impl term::Context) -> anyhow::Result<()> {
+    let profile = ctx.profile()?;
+    let storage = &profile.storage;
+    let signer = term::signer(&profile)?;
+
+    let id = options
+        .id
+        .or_else(|| {
+            radicle::rad::repo(options.path.unwrap_or_else(|| Path::new(".").to_path_buf()))
+                .ok()
+                .map(|(_, id)| id)
+        })
+        .context("Couldn't get ID / Path from command line and cwd is not a reposiitory either")?;
+
+    let project = storage
+        .get(signer.public_key(), id)?
+        .context("No project with such ID exists")?;
+
+    if options.refs {
+        let path = profile
+            .paths()
+            .storage()
+            .join(id.to_human())
+            .join("refs")
+            .join("namespaces");
+
+        Command::new("tree")
+            .current_dir(path)
+            .args(["--noreport", "--prune"])
+            .stdout(Stdio::inherit())
+            .stderr(Stdio::inherit())
+            .spawn()?
+            .wait()?;
+    } else if options.payload {
+        println!(
+            "{}",
+            colorizer().colorize_json_str(&serde_json::to_string_pretty(&project.payload)?)?
+        );
+    } else if options.history {
+        let repo = storage.repository(id)?;
+        let head = Doc::<Untrusted>::head(signer.public_key(), &repo)?;
+        let mut history = repo.revwalk(head)?.collect::<Vec<_>>();
+
+        // Retrieve root document.
+        let oid = history.pop().ok_or(IdentityError::MissingRoot)??;
+        let revision = history.len() as u32;
+
+        let mut counter = 0;
+        print_history(&repo, oid.into(), true, revision == counter)?;
```

Comment:
Why have a separate `print_history` here? Can't we do `print_history(..., counter == 0, counter == history.len())`?

Reply:
We can also generate the encounter with `enumerate()` on the iterator

Reply:
> Why have a separate `print_history` here? Can't we do `print_history(..., counter == 0, counter == history.len())`?

We can actually, just followed the practice [here](https://github.com/radicle-dev/heartwood/blame/master/radicle/src/identity/project.rs#L443-L472).